### ISSUE TITLE
Enable CORS for statuses and inboxes (ActivityPub API)

### ISSUE THREAD

**Issue by wiktor-k** [tootsuite/mastodon#10400]
Hello,

I'd like to ask if it would be possible to enable cross-origin access (CORS) to statuses (`/@user/id`) and inboxes (`/inbox`) just like it's [currently enabled](https://github.com/tootsuite/mastodon/blob/master/config/initializers/cors.rb#L16-L23) for users (`/@user`)

### Pitch

Status endpoint would allow reading statuses from cross-origin resources:
```
    resource '/@:account_username/:id',
      headers: :any,
      methods: [:get],
      credentials: false
```

While inbox endpoint would allow `POST`ing activities:

```
    resource '/inbox',
      headers: :any,
      methods: [:post],
      credentials: false
```

### Motivation

This change allows interacting with Mastodon from ActivityPub clients running entirely in the browser. Currently only [Mastodon specific API](https://github.com/tootsuite/mastodon/blob/master/config/initializers/cors.rb#L24-L28) (`/api/*`) is available from cross-origin pages, this change would allow using pure ActivityPub REST API to interact with the Fediverse.

**Comment by koehn** [comment]
This effectively prevents any C2S application from being able to view objects from Mastodon users. Please consider fixing this. 

https://socialhub.activitypub.rocks/t/cors-restrictions/658

**Comment by wiktor-k** [comment]
@koehn this shouldn't be a big patch, you may try adding something [like this](https://github.com/tootsuite/mastodon/blob/master/config/initializers/cors.rb#L20-L23), testing it locally and submitting a PR here.

I did include also a CORS example for `/inbox`. That'd allow writing ActivityPub clients in pure JavaScript (for example to have account signing keys on the client instead of server, among other things).

**Comment by koehn** [comment]
I'd love to submit a PR for your project, but I'm busy working on my own ActivityPub client and server. 😁 

Right now I need to open a proxy for my web UI to allow users to perform `GET` requests to access objects from remote systems like Mastodon, because of your CORS configuration. That's sub-optimal, to say the least.

**Comment by wiktor-k** [comment]
This is not my project. Actually I created this issue because I also work on my own ActivityPub client and server. It seems everyone and their cat has one nowadays ;)

**Comment by ClearlyClaire** [comment]
#15695 would partially fix this, but as I explained there, I'm reluctant to allow authenticated browser-issued requests.

Indeed, currently, ActivityPub object identifiers are implicitly namespaced by domain name, and any actor on a given domain is considered to be allowed to create any object on that same domain, which means that two users `https://foo.com/users/a` and `https://foo.com/users/b` could both claim to `Create` an object at `https://foo.com/objects/1`.

In a C2S setting, the two user's server is supposed to decide about the object identifier, but if the users themselves hold the signing keys (which would be the case for browser-issued signed requests), they can issue “S2S” queries directly without screening from their home servers.

This means a browser application serving multiple tenants from a same domain cannot possibly be secure with the current security model. An application that would serve only a single user from a given domain name would be fine, though, but I'm not sure it's worth risking having multi-tenant browser-based services be developed only to be running into these security issues.

**Comment by johnspurlock** [comment]
Podcast Index [1] is prototyping out a new rss tag for open-standards-based podcast episode comments, supporting ActivityPub as a transport.

CORS-enabling the Mastodon standard ActivityPub endpoints would be crucial for new web-based podcast apps to display these comments, as these new web-based apps may not have a backend to proxy the requests.

Thanks!

[1] https://github.com/Podcastindex-org/podcast-namespace

**Comment by wiktor-k** [comment]
@johnspurlock there is one PR already solving most of these issues at https://github.com/mastodon/mastodon/pull/15695 but unfortunately it seems it's not a high priority for Mastodon maintainers :shrug:

**Comment by johnspurlock** [comment]
> @johnspurlock there is one PR already solving most of these issues at #15695 but unfortunately it seems it's not a high priority for Mastodon maintainers 🤷

That's too bad.  The proprietary mastodon api supports cors requests for public stuff, but not their ActivityPub endpoints?  Doesn't make sense.

Perhaps at least make it available via a setting?

**Comment by obale** [comment]
Another important endpoint that needs the right CORS configuration is **/outbox** and **/nodeinfo/v2.0** (https://github.com/mastodon/mastodon/issues/23135)

Currently it is possible to get the user's posts via the Mastodon specific API (without authentication) and not via the ActivityPub compliant endpoint. This makes client-side implementation, such as visualizations of timelines on third-party websites highly Mastodon specific.

I am surprised that compatibility with the ActivityPub specification is not higher priority. If it is a resource issue I am happy to jump in, as do many other others, at least according to the created issues and comments around this topic.

The fix seems to be also pretty trivial, except if I am missing something critical. Those endpoints wouldn't expose anything that isn't either way exposed. Furthermore, by introducing a proxy and making those calls server side they are accessible. So no real harm to set the CORS header.

**Comment by wiktor-k** [comment]
> The fix seems to be also pretty trivial, except if I am missing something critical.

Nope you're not missing anything. At least for a wide variety of endpoints the implementation is trivial and has been suggested: https://github.com/mastodon/mastodon/pull/15695.

> I am surprised that compatibility with the ActivityPub specification is not higher priority. 

Unfortunately it's either low-prio or Mastodon maintainers are busy with other things (or both).

> If it is a resource issue I am happy to jump in, as do many other others, at least according to the created issues and comments around this topic.

You can try submitting something similar and see the end-result. Maybe we'll all get lucky and it will finally be merged? I think the vast majority of work won't be coding and testing but rather generating enough Mastodon-social-buzz to get maintainer's attention ;)

**Comment by obale** [comment]
@wiktor-k thank you for the reply. 

As unfortunate it is, that all makes sense. Maybe it would help to collect all the endpoints and requirements to submit a robust proposal.

Additional to the mentioned calls in the original issue description I would like to propose the following additional calls.

```sh
# Important to determine what software the instance is running to adapt the client logic, 
# e.g. have specific calls for Mastodon, Lemmy, Pixelfeed, etc.
GET /nodeinfo/v2.0

# To get all the posts of a specific user. Useful for timelines visualizations on third-party websites
GET /users/:username/outbox

# This is the ID of a post as returned by the /users/:username/outbox call and should be accessible 
# in a programmatic way
GET /users/:username/statuses/:id/activity
```

The most important one, at least in the short term, may be the `/nodeinfo/v2.0`. It allows the client implementation to determine that the called instance is running Mastodon and a fall back to the proprietary API of Mastodon. Not optimal, at least it is workaround that would unblock a lot of implementations.

Are there more?

**Comment by wiktor-k** [comment]
@obale I see you're coming from a different angle. I'm looking at the following APIs:

```sh
# Get specified note by ID. This is what is getting sent around via various ActivityPub endpoints.
GET /users/:username/statuses/:id

# Get user's JSON representation.
GET /users/:username

# Deliver signed note to users on this server.
POST /inbox
```

For the record I'm working on ActivityPub client written in JavaScript that runs only in the browser (no server components). There are of course many, many more interesting endpoints like `followers` but just the bare minimum would make my client usable.

**Comment by obale** [comment]
My comment was meant as addition to the endpoints you mentioned.

For future readers here is the current list we have for CORS changes.

```
# Get specified note by ID. This is what is getting sent around via various ActivityPub endpoints.
GET /users/:username/statuses/:id

# Get user's JSON representation.
GET /users/:username

# Deliver signed note to users on this server.
POST /inbox

# Important to determine what software the instance is running to adapt the client logic, 
# e.g. have specific calls for Mastodon, Lemmy, Pixelfeed, etc.
GET /nodeinfo/v2.0

# To get all the posts of a specific user. Useful for timelines visualizations on third-party websites
GET /users/:username/outbox

# This is the ID of a post as returned by the /users/:username/outbox call and should be accessible 
# in a programmatic way. Not sure what the different to the /activity with the one without is.
GET /users/:username/statuses/:id/activity

# To the replies of a post
GET /users/:username/statuses/:id/replies
```

P.S.: I am currently working on a read-only timeline that uses public posts of an user, so basically a subset you are working on. That is not really relevant for this discussion, because we need all the endpoints fixed to empower the end-user and to assure interoperability and prevent hosting provider lock in.

**Comment by MidnightLightning** [comment]
I filed #31350 as an issue (targeting the "Following" ActivityPub endpoint) before finding this related issue. Commenting here too to bump this issue up in priority and relevance: Mastodon as a project already has CORS implemented in some ActivityPub endpoints, and all the Mastodon-specific API endpoints, so it seems Mastodon as a project is okay with cross-site front-end requests for public data in general. So, it seems odd odd to me that they'd implement this only "halfway" and not be uniform.

https://github.com/mastodon/mastodon/pull/15695 was closed as closed as stale; is it possible for the maintaining team to re-open that pull request and evaluate it and the other ActivityPub endpoints properly handling CORS requests?

**Comment by angelo-v** [comment]
Can we please fix this. I want to build browser apps that access the public ressources, but it is just not possible without CORS being enabled. It is an easy fix, just do it. I will provide an PR if it will get merged and not closed as stale.

**Comment by MidnightLightning** [comment]
@angelo-v It looks like PR https://github.com/mastodon/mastodon/pull/15695 got reopened, though it's now out-of-date enough that it needs a rebase and maybe starting from scratch. But @vmstan has weighed in on one of the duplicate issues, so might have the inertia to get acted-upon this time.

**Comment by wiktor-k** [comment]
> so might have the inertia to get acted-upon this time.

I wouldn't count on that since [it has been closed](https://github.com/mastodon/mastodon/pull/15695#event-14472102802) by @ClearlyClaire with no further commentary.

**Comment by MidnightLightning** [comment]
@wiktor-k Yes, it appears Claire is not interested in updating that pull request to be based on the current codebase, so the path is open for someone else to submit a similar patch for this issue.

**Comment by dolmen** [comment]
There is an important comment by @ClearlyClaire that must not be buried:

https://github.com/mastodon/mastodon/issues/10400#issuecomment-777404527

**Comment by MidnightLightning** [comment]
@dolmen that comment I don't think is that relevant for this specific request, because this request is (a) talking about read-only endpoints (the client making the request wouldn't be authenticated; they're just looking for public info), and (b) specifically the Mastodon client. The situation of User A and User B both having a "Create Object C" message in their "inbox" feed is something that Mastodon should be handling separately (when User A and B attempt to create objects), not when other public clients are just viewing User A and B's feeds.

**Comment by evanp** [comment]
For this particular issue, we have the [proxyUrl](https://www.w3.org/TR/activitypub/#proxyUrl) endpoint defined in ActivityPub. It makes doing cross-server requests really easy, by proxying them through the actor's home server. It really routes around a lot of the CORS issues.

**Comment by wiktor-k** [comment]
> For this particular issue, we have the [proxyUrl](https://www.w3.org/TR/activitypub/#proxyUrl) endpoint defined in ActivityPub.

As far as I can see it's not supported (`git grep proxyUrl` gives zero results in this repository, querying `mastodon.social`'s WebFinger resource for a random account doesn't return it). I don't see a ticket explicitly mentioning this, it's present in [some comments](https://github.com/mastodon/mastodon/issues/24068#issuecomment-1465601687). I think it'd be good to discuss that in a separate ticket.

> It makes doing cross-server requests really easy, by proxying them through the actor's home server. It really routes around a lot of the CORS issues.

Well, if CORS is implemented then there's no need to route around at all.

Note that Mastodon proprietary API's endpoints already support CORS:

```
$ curl -i -H Origin:metacode.biz -sSL https://mastodon.social/api/v1/statuses/114922274092325181/context | grep access-control
access-control-max-age: 7200
access-control-allow-methods: POST, PUT, DELETE, GET, PATCH, OPTIONS
access-control-expose-headers: Link, X-RateLimit-Reset, X-RateLimit-Limit, X-RateLimit-Remaining, X-Request-Id
access-control-allow-origin: *
```
